File: day_3/task.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
)

var directions = [8][2]int{
	{-1, -1}, {-1, 0}, {-1, 1},
	{0, -1}, {0, 1},
	{1, -1}, {1, 0}, {1, 1}}

func readFile(fileName string) ([][]rune, error) {
	file, err := os.Open(fileName)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var inputs [][]rune
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		inputs = append(inputs, []rune(scanner.Text()))
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	return inputs, nil
}

func isSymbol(ch rune) bool {
	return ch != '.' && ch != '\r' && ch != '\n' && !(ch >= '0' && ch <= '9')
}

func isAdjacentToSymbol(schematic [][]rune, row, col int) bool {
	for _, direction := range directions {
		newRow, newCol := row+direction[0], col+direction[1]
		if newRow > 0 && newRow < len(schematic) && newCol > 0 && newCol < len(schematic[newRow]) &&
			isSymbol(schematic[newRow][newCol]) {
			return true
		}
	}
	return false
}

func extractNumber(schematic [][]rune) []int {
	var nums []int
	for i, line := range schematic {
		for j, ch := range line {
			if ch >= '0' && ch <= '9' {
				if j+1 < len(line) && line[j+1] >= '0' && line[j+1] <= '9' {
					num, _ := strconv.Atoi(string([]rune{ch, line[j+1]}))
					if isAdjacentToSymbol(schematic, i, j) || isAdjacentToSymbol(schematic, i, j+1) {
						nums = append(nums, num)
					}
					j++
				} else if isAdjacentToSymbol(schematic, i, j) {
					num, _ := strconv.Atoi(string(ch))
					nums = append(nums, num)
				}
			}
		}
	}
	return nums
}

func part1(schematic [][]rune) (int, error) {
	nums := extractNumber(schematic)
	totalSum := 0
	for _, num := range nums {
		totalSum += num
	}
	return totalSum, nil
}

func main() {
	schematic, err := readFile("day_3/input.txt")
	if err != nil {
		log.Fatal(err)
	}
	totalSum, err := part1(schematic)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println(totalSum)
}

File: day_3/task.py
from collections import defaultdict

filename = "./input.txt"
inputs = []


with open(filename, 'r') as file:
    for input in file.readlines():
        inputs.append(input.strip())


def part1():
    def find_numbers():
        numbers = []
        for i, row in enumerate(inputs):
            j = 0
            while j < len(row):
                if row[j].isdigit():
                    start = j
                    while j < len(row) and row[j].isdigit():
                        j += 1
                    numbers.append((i, start, j))
                else:
                    j += 1
        return numbers

    def get_numbers(i, j1, j2, lenght, input_length):
        number_set = set()
        for j in range(j1, j2):
            for di in [-1, 0, 1]:
                for dj in [-1, 0, 1]:
                    if (di, dj) != (0, 0):
                        nbri, nbrj = i + di, j + dj
                        if 0 <= nbri < lenght and 0 <= nbrj < input_length:
                            number_set.add((nbri, nbrj))
        return number_set

    def is_symbol(char):
        return not char.isdigit() and char != '.'

    def process_numbers(numbers):
        total = 0
        for i, j1, j2 in numbers:
            if any(is_symbol(inputs[nbri][nbrj]) for nbri, nbrj in get_numbers(i, j1, j2, len(inputs), len(inputs[0]))):
                num = int(inputs[i][j1:j2])
                total += num
        return total

    numbers = find_numbers()
    return process_numbers(numbers)


print(part1())

def part2():
    def find_numbers():
        numbers = []
        for i, row in enumerate(inputs):
            j = 0
            while j < len(row):
                if row[j].isdigit():
                    start = j
                    while j < len(row) and row[j].isdigit():
                        j += 1
                    numbers.append((i, start, j))
                else:
                    j += 1
        return numbers

    def get_numbers(i, j1, j2, lenght=len(inputs), input_length=len(inputs[0])):
        number_set = set()
        for j in range(j1, j2):
            for di in [-1, 0, 1]:
                for dj in [-1, 0, 1]:
                    if (di, dj) != (0, 0):
                        nbri, nbrj = i + di, j + dj
                        if 0 <= nbri < lenght and 0 <= nbrj < input_length:
                            number_set.add((nbri, nbrj))
        return number_set

    def is_symbol(char):
        return not char.isdigit() and char != '.'

    def process_numbers(numbers):
        total_sum = 0
        star_numbers_map = defaultdict(list)
        for i, j1, j2 in numbers:
            num = int(inputs[i][j1:j2])
            if any(is_symbol(inputs[nbri][nbrj]) for nbri, nbrj in get_numbers(i, j1, j2)):
                for nbri, nbrj in get_numbers(i, j1, j2):
                    if inputs[nbri][nbrj] == '*':
                        star_numbers_map[(nbri, nbrj)].append(num)

        for x in star_numbers_map.values():
            if len(x) == 2:
                total_sum += (x[0] * x[1])
        return total_sum

    numbers = find_numbers()
    return process_numbers(numbers)


print(part2())
